### PR TITLE
Adds support for building VCRT Forwarders as a UPM package

### PR DESCRIPTION
**Pull request by davidkline-ms**
This change adds a package.json to allow for Unity Package Manager distributtion. Other files that Unity looks for in UPM packages (ex: CHANGELOG.md, etc) are also included.

- [x] Create package.json to define the UPM package
- [x] Add CHANGELOG.md which will be displayed when "View changelog" is clicked
- [x] During pre-packaging, copy CHANGELOG.md, README.md and LICENSE (as LICENSE.md) to the expected locations
- [x] build.cmd create package
- [x] Update README.md to describe the UPM distribution
- [x] Update init cs fie to allow finding the dll in a UPM installation

image

### CODE REVIEW

**Comment by davidkline-ms** [comment]
closing for now as a couple of links are not properly working.... will reopen when fixed.

File: Unity/Editor/InitVCRTForwarders.cs [hunk 1]
```diff
@@ -1,4 +1,5 @@
 using System;
```
Comment (keveleigh) [inline Unity/Editor/InitVCRTForwarders.cs:1]: Missing MSFT license header

File: Unity/package.json [hunk 2]
```diff
@@ -0,0 +1,22 @@
+{
+    "name": "com.microsoft.vcrt.forwarders.140",
```
Comment (keveleigh) [inline Unity/package.json:2]: Will this (and the added metas to this folder) be included in the NuGet package? If so, is that an issue?

File: Unity/CHANGELOG.md.meta [hunk 3]
```diff
@@ -0,0 +1,7 @@
+fileFormatVersion: 2
```
Comment (dfields-msft) [inline Unity/CHANGELOG.md.meta:1]: Why is this meta file included in the changes? (Why does it exist at all?) Likewise for most of the .meta files in this PR.

File: Unity/Editor/InitVCRTForwarders.cs [hunk 4]
```diff
@@ -32,18 +33,37 @@ static InitVCRTForwarders()
             return;
         }
 
+        List<string> searchFolders = new List<string>()
+        {
+            Application.dataPath,                                       // Assets folder
+            Path.GetFullPath(Path.Combine("Library", "PackageCache")),  // Library\PackageCache folder
+            Path.GetFullPath("Packages")                                // Pacakges folder
```
Comment (dfields-msft) [inline Unity/Editor/InitVCRTForwarders.cs:40]: Typo: `// Pacakges folder`
```suggestion
            Path.GetFullPath("Packages")                                // Packages folder
```

File: README.md [hunk 5]
```diff
@@ -11,6 +11,27 @@ on the CRT (debug / release) in use in the component project. For projects using
...
+To install,
+
+- In Unity, open `Edit` > `Project Settings`
+- Select `Package Manager` from the options on the left-hand side of the Project Settings window
+- Add a new Scoped Registry with the following settings
+  - Name: Microsoft Mixed Reality
+  - URL: https://pkgs.dev.azure.com/aipmr/MixedReality-Unity-Packages/_packaging/Unity-packages/npm/registry/
+  - Scopes:
+    - com.microsoft.mixedreality
+    - com.microsoft.spatialaudio
```
Comment (dfields-msft) [inline README.md:27]: These shouldn't be included in generic instructions for including vcrtforwarders. Ideally, there wouldn't be anything MR-specific at all.

```suggestion
    - com.microsoft.mixedreality
    - com.microsoft.spatialaudio
```

File: Unity/Editor/InitVCRTForwarders.cs [hunk 6]
```diff
@@ -32,18 +33,37 @@ static InitVCRTForwarders()
...
+            Application.dataPath,                                       // Assets folder
+            Path.GetFullPath(Path.Combine("Library", "PackageCache")),  // Library\PackageCache folder
+            Path.GetFullPath("Packages")                                // Pacakges folder
+        };
+
         // Find a representative VCRTForwarders binary - there should be only one.
-        var assets = AssetDatabase.FindAssets(Path.GetFileNameWithoutExtension(moduleName));
-        if (assets.Length != 1)
+        string dllDirectory = string.Empty;
+        for (int i = 0; i < searchFolders.Count; i++)
```
Comment (dfields-msft) [inline Unity/Editor/InitVCRTForwarders.cs:45]:
```suggestion
        foreach (var searchFolder in searchFolders)
```

File: Unity/Editor/InitVCRTForwarders.cs [hunk 7]
```diff
@@ -32,18 +33,37 @@ static InitVCRTForwarders()
...
         {
-            Debug.LogError(string.Format("Failed to find single asset for {0}; found {1} instead!", moduleName, assets.Length));
+            string[] files = Directory.GetFiles(searchFolders[i], moduleName, SearchOption.AllDirectories);
```
Comment (dfields-msft) [inline Unity/Editor/InitVCRTForwarders.cs:47]: Assuming you accept the previous suggestion, you'll need this too:
```suggestion
            string[] files = Directory.GetFiles(searchFolder, moduleName, SearchOption.AllDirectories);
```

File: Unity/Editor/InitVCRTForwarders.cs [hunk 8]
```diff
@@ -32,18 +33,37 @@ static InitVCRTForwarders()
             return;
         }
 
+        List<string> searchFolders = new List<string>()
```
Comment (dfields-msft) [inline Unity/Editor/InitVCRTForwarders.cs:36]: Do these logic changes apply equally to the nuget distribution? @majaeger FYI.

File: Unity/package.json (hunk 2 above)
Comment (dfields-msft) [reply Unity/package.json:2]: Definitely need to minimize impact on the nuget. Long term, we might want to actually remove the Unity support from the NuGet (that's a breaking change, so we'd need to bump the major version...).

Comment (davidkline-ms) [reply Unity/package.json:2]: good question. i will investigate

File: vcrt-forwarders.nuspec [hunk 9]
```diff
@@ -13,6 +13,9 @@
         <requireLicenseAcceptance>true</requireLicenseAcceptance>
         <description>App-local DLLs that forward to runtime components of Visual C++ libraries which are required to run Visual C++ applications. The runtime components being forwarded to should already be installed on a given machine, this can be done using the Visual C++ Redistributable Packages https://aka.ms/AA4pp63. See https://github.com/Microsoft/vcrt-forwarders for more details. </description>
         <releaseNotes>
+            1.0.8
+            Add support for installing in the unity package Cache
```
Comment (dfields-msft) [inline vcrt-forwarders.nuspec:17]: From the nuget perspective, this isn't really a material change; in a perfect world, this PR would actually not affect the nuget at all. The only reason there's any impact at all is that the .cs script is shared between nuget and UPM, so this comment should be something benign about "revised DLL path discovery logic in Unity initialization script"

**Review by dfields-msft — CHANGES_REQUESTED** [review]
Definitely need to touch base with @manodasanW on the strategy for this change, as well as necessary build pipeline changes to actually produce and publish the UPM package.

File: Unity/CHANGELOG.md.meta (hunk 3 above)
Comment (davidkline-ms) [reply Unity/CHANGELOG.md.meta:1]: it is required by unity. if it isnt present, the file will not be acknowledged. the .md file will be placed next to the meta file

**Comment by davidkline-ms** [comment]
Closing (again) while some non-PR issues are resolved.

**Pull request closed without merging**